zip-0244.rst

```diff
@@ -1,8 +1,382 @@
+specified by the ZIP 307 Light Client Protocol [#zip-0307]_.
+
+This digest is a BLAKE2b-256 hash of the following values ::
+
+   * 3a. ``sapling_spends_digest`` (32-byte hash)
+   * 3b. ``sapling_outputs_digest`` (32-byte hash)
+   * 3c. ``valueBalance`` (64-bit signed little-endian)
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSaplingHash"
+
+3a: ``sapling_spends_digest``
+''''''''''''''''''''''''''''''
+This digest is a BLAKE2b-256 hash of the following values ::
+
+   * 3a.i.   ``sapling_spends_compact_digest`` (32-byte hash)
+   * 3b.ii.  ``sapling_spends_noncompact_digest`` (32-byte hash)
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSSpendsHash"
+
+3a.i: ``sapling_spends_compact_digest``
+.......................................
+A BLAKE2b-256 hash of the field encoding of all nullifier field
+values of Sapling shielded spends belonging to the transaction.
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSSpendCHash"
+
+3a.ii: ``sapling_spends_noncompact_digest``
+...........................................
+A BLAKE2b-256 hash of the non-nullifier information for all Sapling shielded spends
+belonging to the transaction. For each spend, the following elements are included
+in the hash:: 
+
+   * 3a.ii.1 ``cv``      (field encoding bytes)
+   * 3a.ii.2 ``anchor``  (field encoding bytes)
+   * 3a.ii.3 ``rk``      (field encoding bytes)
+   * 3a.ii.4 ``zkproof`` (field encoding bytes)
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSSpendNHash"
+
+3b: ``sapling_outputs_digest``
+'''''''''''''''''''''''''''''''
+This digest is a BLAKE2b-256 hash of the following values ::
+
+   * 3a.i.   ``sapling_outputs_compact_digest`` (32-byte hash)
+   * 3b.ii.  ``sapling_outputs_memos_digest`` (32-byte hash)
+   * 3b.iii. ``sapling_outputs_noncompact_digest`` (32-byte hash)
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSOutputHash"
+
+3b.i: ``sapling_outputs_compact_digest``
+.........................................
+A BLAKE2b-256 hash of the subset of Sapling output information included in the 
+ZIP-307 ``CompactBlock`` format for all Sapling shielded outputs
+belonging to the transaction. For each output, the following elements are included
+in the hash:: 
+
+   * 3b.i.1 ``cmu``                  (field encoding bytes)
+   * 3b.i.2 ``ephemeral_key``        (field encoding bytes)
+   * 3b.i.3 ``enc_ciphertext[..52]`` (First 52 bytes of field encoding)
+
+The personalization field of this hash is set to::
+
+  "ZTxIdSOutC__Hash"
+
+3a.ii: ``sapling_outputs_memos_digest``
+........................................
+A BLAKE2b-256 hash of the subset of Sapling shielded memo field data for all Sapling 
+shielded outputs belonging to the transaction. For each output, the following elements 
+are included in the hash:: 
+
+   * 3b.ii.1 ``enc_ciphertext[52..564] (contents of the encrypted memo field)
```

Comment:
This doesn't include the authentication tag. But the light client should check the tag on the entire ciphertext, if it needs the memo. (It can check the first 52 bytes via the note commitment, but it can't do that for the memo bytes.)

Reply:
@str4d when the light clients download the memo data, are they actually downloading the whole `enc_ciphertext`?

Reply:
Here's a thought: does keeping it the way it's written allow light clients to prune stored data, but keep an encrypted version of the memo in a way that allows them to later drop the remainder of the "noncompact" data, but use the Merkle path to verify the integrity of the memo? Or, could the lightwallet server provide the Merkle path along with the pruned memo data instead of providing the remainder of the enc_ciphertext?